File: app/src/main/java/com/tobiasfried/brewkeeper/HistoryFragment.java
package com.tobiasfried.brewkeeper;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
import com.firebase.ui.firestore.FirestoreRecyclerOptions;
import com.google.android.material.button.MaterialButton;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.tobiasfried.brewkeeper.model.Brew;
import com.tobiasfried.brewkeeper.utils.TimeUtility;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static com.tobiasfried.brewkeeper.EntryActivity.EXTRA_BREW_ID_HISTORY;

public class HistoryFragment extends Fragment {

    private static final String LOG_TAG = HistoryFragment.class.getSimpleName();

    private FirebaseFirestore db;
    private FirestoreRecyclerAdapter<Brew, HistoryViewHolder> mAdapter;
    private FirestoreRecyclerOptions<Brew> options;
    private String sortOptions;
    private Query.Direction sortOrder = Query.Direction.ASCENDING;
    private Unbinder unbinder;

    private View rootView;

    @BindView(R.id.list)
    RecyclerView recyclerView;

    @BindView(R.id.spinner_sort_by)
    Spinner sortSpinner;

    @BindView(R.id.button_sort_order)
    MaterialButton sortOrderButton;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Get Database instance
        db = FirebaseFirestore.getInstance();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        rootView = inflater.inflate(R.layout.fragment_brews, container, false);
        unbinder = ButterKnife.bind(this, rootView);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));

        // Set Spinner Adapter
        ArrayAdapter<CharSequence> sortAdapter = ArrayAdapter.createFromResource(Objects.requireNonNull(getContext()),
                R.array.array_sort_names_history, R.layout.spinner_item_sort);
        sortSpinner.setAdapter(sortAdapter);
        sortSpinner.setSelection(0);
        sortSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                sortOptions = getResources().getStringArray(R.array.array_sort_options)[position];
                setupRecyclerView();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        // Set Sort Order Button
        sortOrderButton.setOnClickListener(v -> {
            float rotation = v.getRotation();
            v.setRotation(rotation == 0 ? 180 : 0);
            sortOrder = rotation == 0 ? Query.Direction.ASCENDING : Query.Direction.DESCENDING;
            setupRecyclerView();
        });

        return rootView;
    }

    private void setupRecyclerView() {
        // Inflate and setup RecyclerView
        Query query = db.collection(Brew.HISTORY)
                .orderBy(sortOptions, sortOrder);
        FirestoreRecyclerOptions<Brew> options = new FirestoreRecyclerOptions.Builder<Brew>()
                .setQuery(query, Brew.class)
                .setLifecycleOwner(this)
                .build();
        mAdapter = new FirestoreRecyclerAdapter<Brew, HistoryViewHolder>(options) {
            @NonNull
            @Override
            public HistoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                View itemView = getLayoutInflater().inflate(R.layout.list_item_history, parent, false);
                return new HistoryViewHolder(itemView);
            }

            @Override
            protected void onBindViewHolder(@NonNull HistoryViewHolder holder, int position, @NonNull Brew brew) {
                // Apply fields
                holder.name.setText(brew.getRecipe().getName());
                holder.date.setText(TimeUtility.formatDateShort(brew.getEndDate()));

                // Set ClickListener
                final String brewId = getSnapshots().getSnapshot(position).getId();
                holder.card.setOnClickListener(v -> {
                    Intent intent = new Intent(getActivity(), EntryActivity.class);
                    intent.putExtra(EXTRA_BREW_ID_HISTORY, brewId);
                    startActivity(intent);
                });
            }
        };

        recyclerView.setAdapter(mAdapter);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }
}

File: app/src/main/java/com/tobiasfried/rasa/history/HistoryFragment.kt
package com.tobiasfried.rasa.history

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner

import com.firebase.ui.firestore.FirestoreRecyclerAdapter
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import com.google.android.material.button.MaterialButton
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.tobiasfried.rasa.domain.Brew
import com.tobiasfried.rasa.utils.TimeUtility

import java.util.Objects
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.Unbinder
import com.tobiasfried.rasa.DetailActivity
import com.tobiasfried.rasa.R
import com.tobiasfried.rasa.entry.EntryActivity.Companion.EXTRA_BREW_ID_HISTORY

class HistoryFragment : Fragment() {

    private var db: FirebaseFirestore? = null
    private var mAdapter: FirestoreRecyclerAdapter<Brew, HistoryViewHolder>? = null
    private val options: FirestoreRecyclerOptions<Brew>? = null
    private var sortOptions: String? = null
    private var sortOrder: Query.Direction = Query.Direction.DESCENDING
    private var unbinder: Unbinder? = null

    private var rootView: View? = null

    @BindView(R.id.list)
    internal var recyclerView: RecyclerView? = null

    @BindView(R.id.spinner_sort_by)
    internal var sortSpinner: Spinner? = null

    @BindView(R.id.button_sort_order)
    internal var sortOrderButton: MaterialButton? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Get Database instance
        db = FirebaseFirestore.getInstance()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        rootView = inflater.inflate(R.layout.fragment_history, container, false)
        unbinder = ButterKnife.bind(this, rootView!!)
        recyclerView!!.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)

        // Set Spinner Adapter
        val sortAdapter = ArrayAdapter.createFromResource(Objects.requireNonNull<Context>(context),
                R.array.array_sort_names_history, R.layout.spinner_item_sort)
        sortSpinner!!.adapter = sortAdapter
        sortSpinner!!.setSelection(0)
        sortSpinner!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                sortOptions = resources.getStringArray(R.array.array_sort_options)[position]
                setupRecyclerView()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        // Set Sort Order Button
        sortOrderButton!!.setOnClickListener { v ->
            val rotation = v.rotation
            v.rotation = (if (rotation == 0f) 180 else 0).toFloat()
            sortOrder = if (rotation == 0f) Query.Direction.ASCENDING else Query.Direction.DESCENDING
            setupRecyclerView()
        }

        return rootView
    }

    private fun setupRecyclerView() {
        // Inflate and setup RecyclerView
        val query = db!!.collection(Brew.HISTORY)
                .orderBy(sortOptions!!, sortOrder)
        val options = FirestoreRecyclerOptions.Builder<Brew>()
                .setQuery(query, Brew::class.java)
                .setLifecycleOwner(this)
                .build()
        mAdapter = object : FirestoreRecyclerAdapter<Brew, HistoryViewHolder>(options) {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HistoryViewHolder {
                val itemView = layoutInflater.inflate(R.layout.list_item_history, parent, false)
                return HistoryViewHolder(itemView)
            }

            override fun onBindViewHolder(holder: HistoryViewHolder, position: Int, brew: Brew) {
                // Apply fields
                holder.name!!.text = brew.recipe?.name
                holder.date!!.text = TimeUtility.formatDateShort(brew.endDate)

                // Set ClickListener
                val brewId = snapshots.getSnapshot(position).id
                holder.card!!.setOnClickListener { v ->
                    val intent = Intent(activity, DetailActivity::class.java)
                    intent.putExtra(EXTRA_BREW_ID_HISTORY, brewId)
                    startActivity(intent)
                }
            }
        }

        recyclerView!!.adapter = mAdapter
    }

    override fun onDestroyView() {
        super.onDestroyView()
        unbinder!!.unbind()
    }

}
